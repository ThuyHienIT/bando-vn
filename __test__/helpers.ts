import { randomUUID } from 'crypto';

export function generateFacility(
  type: FacilityTypeEnum,
  name: string = 'Venus Meeting Room'
): FacilityItem {
  return {
    phoneNumber: '24709620',
    id: randomUUID(),
    address: '1 JALAN KILANG,S159402',
    description:
      'Current majority suffer hair born.\nReport daughter government item keep political cut. Customer apply couple. Data happen hear must almost.',
    name,
    amenities:
      'Friend ok nation international line another within.\nUp rate when sometimes probably measure with. Product town sort. Include even alone certain.',
    type: type,
    operationHours: ['8:00', '19:00'],
    offDays: ['2022-04-23T23:18:53Z'],
    thumbnail: 'https://picsum.photos/400/200',
  };
}

export function generateBooking(facilityId: string): BookingItem {
  return {
    facilityId: facilityId,
    from: '2023-04-23T10:30:00Z',
    to: '2023-04-23T11:00:00Z',
    userEmail: 'kqthang1505@gmail.com',
    id: randomUUID(),
    createdAt: '2023-02-18T22:25:56+08:00',
    updatedAt: '2023-02-18T22:25:56+08:00',
  };
}
