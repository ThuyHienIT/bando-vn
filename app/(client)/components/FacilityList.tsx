'use client';

import Image from 'next/image';
// import { memo } from 'react';
// import { Card, Typography } from 'antd';
import { FacilityItemCmp } from '@app/app/(client)/components/FacilityItem';
import { Col, Row, Typography } from 'antd';
import styled from 'styled-components';
import Link from 'next/link';
import { useMemo } from 'react';
import { DoubleRightOutlined } from '@ant-design/icons';

const ColStyle = styled(Col)``;

const ListStyle = styled(Row)`
  ${ColStyle} {
    margin-bottom: 16px;
  }
`;

const LinkRow = styled.div`
  text-align: right;
`;

interface Props {
  data: FacilityItem[];
  heading: string;
  max?: number;
  viewAllLink?: string;
  className?: string;
}

export function FacilityList(props: Props) {
  const listData = useMemo(() => {
    if (!props.max) return props.data;

    return props.data.slice(0, props.max);
  }, [props.data, props.max]);

  return (
    <div className={props.className}>
      <Typography.Title level={4}>{props.heading}</Typography.Title>
      <ListStyle gutter={16}>
        {listData.map((item) => (
          <ColStyle key={item.id} md={6}>
            <FacilityItemCmp data={item} />
          </ColStyle>
        ))}
      </ListStyle>
      {props.max && props.data.length > props.max && props.viewAllLink && (
        <LinkRow>
          <Link href={props.viewAllLink}>
            View all <DoubleRightOutlined />
          </Link>
        </LinkRow>
      )}
    </div>
  );
}
