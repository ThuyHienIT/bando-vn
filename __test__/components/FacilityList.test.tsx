import '@testing-library/jest-dom';

import { generateFacility } from '__test__/helpers';

import { FacilityList } from '@components/FacilityList';
import { FacilityTypeEnum } from '@enums';
import { render, screen } from '@testing-library/react';

describe('Facility List', () => {
  it('less than max', () => {
    const fac = generateFacility(FacilityTypeEnum.Facility);
    fac.id = 'test-id';
    const { container } = render(
      <FacilityList heading="Facilities" max={2} data={[fac]} />
    );

    const heading = screen.getByRole('heading', { name: 'Facilities' });
    const link = screen.queryByTestId('view-all');

    expect(heading).toBeInTheDocument();
    expect(link).toBeNull();

    expect(container).toMatchSnapshot();
  });

  it('more than max', () => {
    const fac1 = generateFacility(FacilityTypeEnum.Facility);
    const fac2 = generateFacility(FacilityTypeEnum.Facility);
    const fac3 = generateFacility(FacilityTypeEnum.Facility);

    fac1.id = 'test-id-1';
    fac2.id = 'test-id-2';
    fac3.id = 'test-id-3';

    const { container } = render(
      <FacilityList
        heading="Facilities"
        max={2}
        data={[fac1, fac2, fac3]}
        viewAllLink="/facilities"
      />
    );

    const link = screen.getByTestId('view-all');

    expect(link).toBeInTheDocument();
    expect(link.getAttribute('href')).toBe('/facilities');

    expect(container).toMatchSnapshot();
  });
});
