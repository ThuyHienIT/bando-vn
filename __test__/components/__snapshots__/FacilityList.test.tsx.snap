// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Facility List less than max 1`] = `
.c1 {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1.c1 .ant-card-actions {
  margin-top: auto;
}

.c2 {
  width: auto;
  height: auto;
}

.c0 {
  margin-top: 32px;
}

<div>
  <div
    class="c0"
  >
    <h4
      class="ant-typography css-dev-only-do-not-override-diro6f"
    >
      Facilities
    </h4>
    <div
      class="ant-row css-dev-only-do-not-override-diro6f"
      style="margin-left: -8px; margin-right: -8px;"
    >
      <div
        class="ant-col ant-col-md-6 css-dev-only-do-not-override-diro6f"
        style="padding-left: 8px; padding-right: 8px;"
      >
        <div
          class="ant-card ant-card-bordered ant-card-small c1 css-dev-only-do-not-override-diro6f"
        >
          <div
            class="ant-card-cover"
          >
            <img
              alt="Venus Meeting Room"
              class="c2"
              data-nimg="1"
              decoding="async"
              height="100"
              loading="lazy"
              src="/_next/image?url=https%3A%2F%2Fpicsum.photos%2F400%2F200&w=640&q=75"
              srcset="/_next/image?url=https%3A%2F%2Fpicsum.photos%2F400%2F200&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fpicsum.photos%2F400%2F200&w=640&q=75 2x"
              style="color: transparent;"
              width="240"
            />
          </div>
          <div
            class="ant-card-body"
          >
            <h4
              class="ant-typography css-dev-only-do-not-override-diro6f"
            >
              Venus Meeting Room
            </h4>
            <span
              class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line css-dev-only-do-not-override-diro6f"
            >
              Current majority suffer hair born.
Report daughter government item keep political cut. Customer apply couple. Data happen hear must almost.
            </span>
          </div>
          <ul
            class="ant-card-actions"
          >
            <li
              style="width: 100%;"
            >
              <span>
                <a
                  href="/facility/test-id"
                >
                  <button
                    class="ant-btn css-dev-only-do-not-override-diro6f ant-btn-text ant-btn-sm"
                    type="button"
                  >
                    <span
                      aria-label="eye"
                      class="anticon anticon-eye"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="eye"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M942.2 486.2C847.4 286.5 704.1 186 512 186c-192.2 0-335.4 100.5-430.2 300.3a60.3 60.3 0 000 51.5C176.6 737.5 319.9 838 512 838c192.2 0 335.4-100.5 430.2-300.3 7.7-16.2 7.7-35 0-51.5zM512 766c-161.3 0-279.4-81.8-362.7-254C232.6 339.8 350.7 258 512 258c161.3 0 279.4 81.8 362.7 254C791.5 684.2 673.4 766 512 766zm-4-430c-97.2 0-176 78.8-176 176s78.8 176 176 176 176-78.8 176-176-78.8-176-176-176zm0 288c-61.9 0-112-50.1-112-112s50.1-112 112-112 112 50.1 112 112-50.1 112-112 112z"
                        />
                      </svg>
                    </span>
                    <span>
                      Details
                    </span>
                  </button>
                </a>
              </span>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Facility List more than max 1`] = `
.c1 {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1.c1 .ant-card-actions {
  margin-top: auto;
}

.c2 {
  width: auto;
  height: auto;
}

.c3 {
  text-align: right;
}

.c0 {
  margin-top: 32px;
}

<div>
  <div
    class="c0"
  >
    <h4
      class="ant-typography css-dev-only-do-not-override-diro6f"
    >
      Facilities
    </h4>
    <div
      class="ant-row css-dev-only-do-not-override-diro6f"
      style="margin-left: -8px; margin-right: -8px;"
    >
      <div
        class="ant-col ant-col-md-6 css-dev-only-do-not-override-diro6f"
        style="padding-left: 8px; padding-right: 8px;"
      >
        <div
          class="ant-card ant-card-bordered ant-card-small c1 css-dev-only-do-not-override-diro6f"
        >
          <div
            class="ant-card-cover"
          >
            <img
              alt="Venus Meeting Room"
              class="c2"
              data-nimg="1"
              decoding="async"
              height="100"
              loading="lazy"
              src="/_next/image?url=https%3A%2F%2Fpicsum.photos%2F400%2F200&w=640&q=75"
              srcset="/_next/image?url=https%3A%2F%2Fpicsum.photos%2F400%2F200&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fpicsum.photos%2F400%2F200&w=640&q=75 2x"
              style="color: transparent;"
              width="240"
            />
          </div>
          <div
            class="ant-card-body"
          >
            <h4
              class="ant-typography css-dev-only-do-not-override-diro6f"
            >
              Venus Meeting Room
            </h4>
            <span
              class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line css-dev-only-do-not-override-diro6f"
            >
              Current majority suffer hair born.
Report daughter government item keep political cut. Customer apply couple. Data happen hear must almost.
            </span>
          </div>
          <ul
            class="ant-card-actions"
          >
            <li
              style="width: 100%;"
            >
              <span>
                <a
                  href="/facility/test-id-1"
                >
                  <button
                    class="ant-btn css-dev-only-do-not-override-diro6f ant-btn-text ant-btn-sm"
                    type="button"
                  >
                    <span
                      aria-label="eye"
                      class="anticon anticon-eye"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="eye"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M942.2 486.2C847.4 286.5 704.1 186 512 186c-192.2 0-335.4 100.5-430.2 300.3a60.3 60.3 0 000 51.5C176.6 737.5 319.9 838 512 838c192.2 0 335.4-100.5 430.2-300.3 7.7-16.2 7.7-35 0-51.5zM512 766c-161.3 0-279.4-81.8-362.7-254C232.6 339.8 350.7 258 512 258c161.3 0 279.4 81.8 362.7 254C791.5 684.2 673.4 766 512 766zm-4-430c-97.2 0-176 78.8-176 176s78.8 176 176 176 176-78.8 176-176-78.8-176-176-176zm0 288c-61.9 0-112-50.1-112-112s50.1-112 112-112 112 50.1 112 112-50.1 112-112 112z"
                        />
                      </svg>
                    </span>
                    <span>
                      Details
                    </span>
                  </button>
                </a>
              </span>
            </li>
          </ul>
        </div>
      </div>
      <div
        class="ant-col ant-col-md-6 css-dev-only-do-not-override-diro6f"
        style="padding-left: 8px; padding-right: 8px;"
      >
        <div
          class="ant-card ant-card-bordered ant-card-small c1 css-dev-only-do-not-override-diro6f"
        >
          <div
            class="ant-card-cover"
          >
            <img
              alt="Venus Meeting Room"
              class="c2"
              data-nimg="1"
              decoding="async"
              height="100"
              loading="lazy"
              src="/_next/image?url=https%3A%2F%2Fpicsum.photos%2F400%2F200&w=640&q=75"
              srcset="/_next/image?url=https%3A%2F%2Fpicsum.photos%2F400%2F200&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fpicsum.photos%2F400%2F200&w=640&q=75 2x"
              style="color: transparent;"
              width="240"
            />
          </div>
          <div
            class="ant-card-body"
          >
            <h4
              class="ant-typography css-dev-only-do-not-override-diro6f"
            >
              Venus Meeting Room
            </h4>
            <span
              class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line css-dev-only-do-not-override-diro6f"
            >
              Current majority suffer hair born.
Report daughter government item keep political cut. Customer apply couple. Data happen hear must almost.
            </span>
          </div>
          <ul
            class="ant-card-actions"
          >
            <li
              style="width: 100%;"
            >
              <span>
                <a
                  href="/facility/test-id-2"
                >
                  <button
                    class="ant-btn css-dev-only-do-not-override-diro6f ant-btn-text ant-btn-sm"
                    type="button"
                  >
                    <span
                      aria-label="eye"
                      class="anticon anticon-eye"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="eye"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M942.2 486.2C847.4 286.5 704.1 186 512 186c-192.2 0-335.4 100.5-430.2 300.3a60.3 60.3 0 000 51.5C176.6 737.5 319.9 838 512 838c192.2 0 335.4-100.5 430.2-300.3 7.7-16.2 7.7-35 0-51.5zM512 766c-161.3 0-279.4-81.8-362.7-254C232.6 339.8 350.7 258 512 258c161.3 0 279.4 81.8 362.7 254C791.5 684.2 673.4 766 512 766zm-4-430c-97.2 0-176 78.8-176 176s78.8 176 176 176 176-78.8 176-176-78.8-176-176-176zm0 288c-61.9 0-112-50.1-112-112s50.1-112 112-112 112 50.1 112 112-50.1 112-112 112z"
                        />
                      </svg>
                    </span>
                    <span>
                      Details
                    </span>
                  </button>
                </a>
              </span>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <div
      class="c3"
    >
      <a
        data-testid="view-all"
        href="/facilities"
      >
        View all 
        <span
          aria-label="double-right"
          class="anticon anticon-double-right"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="double-right"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M533.2 492.3L277.9 166.1c-3-3.9-7.7-6.1-12.6-6.1H188c-6.7 0-10.4 7.7-6.3 12.9L447.1 512 181.7 851.1A7.98 7.98 0 00188 864h77.3c4.9 0 9.6-2.3 12.6-6.1l255.3-326.1c9.1-11.7 9.1-27.9 0-39.5zm304 0L581.9 166.1c-3-3.9-7.7-6.1-12.6-6.1H492c-6.7 0-10.4 7.7-6.3 12.9L751.1 512 485.7 851.1A7.98 7.98 0 00492 864h77.3c4.9 0 9.6-2.3 12.6-6.1l255.3-326.1c9.1-11.7 9.1-27.9 0-39.5z"
            />
          </svg>
        </span>
      </a>
    </div>
  </div>
</div>
`;
